README.md
# wlanpoke
Mitigate Wireless Connectivity Loss in Squeezebox Radios

Version 0.8.7.1a

Introduction

The wlanpoke software attempts to mitigate the "Wireless Connectivity Loss" issues suffered by the current version Squeezebox Baby radio software. The software periodically tests the wireless network, and if it fails for a while, restarts the wireless system without rebooting. The resulting network outage is under a minute, which may or may not interrupt music playback for a time or until resumed.

Connected ssh sessions frequently stay open during a restart.

The software sends failure incident report 'logs' to a computer running netcat (nc or ncat) as a server or a similar tcp logger. This computer should have enough disk space to accommodate log files. One or multiple servers on one or more machines can be used, and the software is configured to report to one of them.

Besides mitigating the wireless connectivity issue, the software instruments the radio's wireless behavior, and this data may be useful to developers troubleshooting the issues. However, you may not want that data collected and/or preserved. Fortunately, affirmative steps are required to setup and enable data logging. Also, the software has the option of disabling logging.

The software is implemented as the shell script "wlanpoke.sh." plus supporting files and documentation. There is also an optional script simple 'web server' to serve statistics to a web browser.

This software has been rushed together in response to an emergency of constantly failing radios. Versions have been keeping the author's 4 SB and 3 UE radios connected for 6 months as of this writing. Several users have downloaded previous versions, and they have reported that it has generally worked to keep their radios usable. This software is just a temporary partial mitigation of a serious reliability issue that has arisen during 2020. There are many un- or under-tested functions and features, multitudes of bugs, questionable methods, and coding ugliness. Please assist in pointing these out or fixing them.

We hope this software will become obsolete when the root causes of the unreliability have been fixed. In the mean time, your kind suggestions and improvements are appreciated. Please feel free to modify and adapt the scripts for your own purposes. And please feel free to join the effort to find real lasting solutions to this issue.

Please address your responses to the slimdevices forums  https://forums.slimdevices.com/showthread.php?109953-WiFi-connection-unstable-lost-on-three-Radios or https://forums.slimdevices.com/showthread.php?111663-Community-Build-Radio-Firmware

Use the latter for more technical questions and discussions.

NOTE: Experimental Failure Reduction

This version is an experimental release that incorporates a brute force application of an experimental method to reduce WLAN failure. Some portable or mobile functionality may be degraded. You may wish to revert your system to a previous version. Check for updates. Read history.txt. Please post your experiences on the Squeezebox Radio forum and the GitHub issues tab.

A possible cause for failure has been identified as too frequent scanning for alternate access points by the radio even when connected to a strong signal router or access point. This scanning may encounter signals or packets that cause the radio to lose wireless sensitivity, degrade reception, and eventually fail. This version includes a function to attempt to reduce the number of these scans. The function uses very poorly or undocumented calls to the radio's wireless driver and chip in the belief that this will have the desired effect. Testing over a few days suggests that this provides a dramatic improvement in some fixed location radios. Other fixed location radios show less improvement. However, other use cases, especially ones that require frequent switching wireless networks, may be degraded. 

The function is controlled by a small text file in the install directory (/etc/wlanpoke) named "scanctrl.inc." It contains at minimum 6 numbers that disable or enable various aspects of the radio's scan function under some circumstances. The default is a list of 6 zeros, to disable everything possible with this function. That may be too much for your use case.

If your experience is degraded, you may wish to disable this function altogether. To do so, overwrite any existing "scanctrl.inc." file with a short (< 11 character) message, e.g.,
  echo "stop" > scanctrl.inc
will do it. Or you can delete the file and then "touch scanctrl.inc" to create a zero length file. If you delete the file, it will be recreated with default values the next time wlanpoke launches.
You can experiment with the parameters on the fly, e.g., 
  echo "0 0 1 0 0 0" > scanctrl.inc			(a nonsense example)
They will be applied after the next quick or full reset. You can pore over the documentation by reading the script and entering /lib/atheros/wmiconfig to get its documentation as it were.

News:

Version 0.8.7.1 (experimental) see above.

Version 0.8.6 (unreleased) fixed a bug that crashed the script when an access point name contains an apostrophe

Version 0.8.5 enables the quick reset by default and improves the RawFails Resets report to show the effect of the quick reset more clearly.

Version 0.8.4 adds two measures of network outages. A Gaps report lists recent outages as gaps and connected times in real seconds. A Resets report similarly lists recent outages resulting in resets. A unix time stamp has been added to the launch time and some other messages, and is used to accurately calculate elapsed seconds. A bug that produced a mostly blank status web page and RawFails report has been fixed.

Version 0.8.3 includes a new function calculates a safe full reset hold time based upon the latest full reset result to better handle lengthy network outages. Timestamps are reported without the time zone (e.g., -0700) suffix, and with an underbar replacing the 'T'. The full reset no longer causes a wpa_cli changed PID report. A bug that caused an incorrect 0 full reset recovery result has been fixed. RawFails report and startup log entries now show the launch options.

Version 0.8.2 was a very large revision of the software to address issues with low signal levels in the presence of interference causing the radio's poor wireless reception, which resulted in long delays after a reset to reestablish a connection. The prior software's fixed trial limit before reset (6) would trigger another full reset before the connection could be reestablished. The limit is now specified in seconds (default 12) and increased in steps if a lower limit fails. The "failed pings" counter keeps incrementing for a better measurement of outage time. Additional logging has been added to evaluate the system's performance in new ways. In particular a new reset recovery by limit time report shows the last several recovery times after a full reset indexed by the reset limit. The software now handles switching between Ethernet and wireless connections: an Ethernet connection disables wireless resets, although ping failures are still monitored and logged. Other code changes include improved math calculations, and a start at more meaningful and conforming variable names.

The investigation into the root causes of the outages continues. The wireless monitoring utility wpa_cli is under suspician. Code has been added to test for its presense and to relaunch it if it is no longer running. This utility working is important to the quick reset method. The full reset method restarts the entire network stack, including wpa_cli, so it is not affected. You may wish to edit the ResetQuick function to add or remove commands as the investigation continues, or just disable it if it the script does not *always* work on your system. Starting in 0.7.6, the quick reset method is disabled by default.

Note: the GitHub download system creates a zip file containing a folder with the same name as the zip file name. This is no longer a flat file, and, when unzipped, deposits the software into a sub folder named for the version. The contents of this sub folder must be copied to the installation folder, which is a step that was not described in the original manual.txt instructions. New versions of this file, with higher minor numbers, are placed in the main branch as they are written, and may be helpful. Good luck!


history.txt
                    wlanpoke To Do and History


To Do:

2022-12-02.02	Enh: new quick method to reload the chip firmware without reloading the driver, 
					keeping the wpa_cli, udhcp, etc., still running, saving time...

2021-11-12.01	Bug: script exits with shell error
					/etc/wlanpoke/wlanpoke.sh: line 1: can't fork
					/etc/wlanpoke/wlanpoke.sh: line 1: can't fork
					sh: 8: unknown operand
					...
					sh: 8: unknown operand
					/etc/wlanpoke/wlanpoke.sh: line 1: can't fork
					...
					/etc/wlanpoke/wlanpoke.sh: line 1: can't fork
					wlanpoke exiting
					/etc/wlanpoke/wlanpoke.sh: line 1: can't fork
					...
					2021-11-07_22:10:25 MuffyTunes Radio 4.27_0841 wlanpoke 754 exiting: Qr:0 Fr:6 Wr:20 Wc:0 [ 379145 1383 48 1 0 0 0 0 14:1 27:1 46:1 ]
				Try using newly 'discovered' ash -x switch to capture debug output. Console or shell script:
					TCPPORT="1122"          # 1121 is for wlanpoke
					TCPLOG=$SERVER			# replace this with ip address of ncat listening on 1122
					WLPIPENAME="tester"
					rm -f $WLPIPENAME
					mkfifo $WLPIPENAME
					while true; do cat $WLPIPENAME | nc -w 3 $TCPLOG $TCPPORT ; done &
					ash -x /etc/wlanpoke/wlanpoke.sh -x -W quick -d /etc/wlanpoke/log/ >$WLPIPENAME 2>&1 &
				The wlanpoke load with the tracing is 13%, pretty high, but the radio still works. We'll see what happens when the network fails! Let's hope the pipe fills until the connection next succeeds and empties the pipe.


2021-04-28.04	Enh: prevent nc hold off (e.g., from ncat) from hanging the script.

2021-04-17.01	Enh: check and relaunch the 'slow' 'web' server if not running.

2021-04-15.03	Enh: quiet the Squeezeplay logging so that messages is more relevant to outages.
					-> need step by step for this.

2021-04-15.02	Enh: add counter for gaps in the music - 
					instrument jive, or perhaps jive_alsa.
					watch 'messages'?

2021-04-15.01	Enh: some sort of statistic to evaluate quick reset so we can evaluate different quick reset methods.
					Cannot just use score, since multiple quick reset short-lived recoveries ending in a full reset may not be desirable.
					How many unsuccessful trials (full reset first time use). Not quick reset's fault?
					  -- easier: increment successful trials ping after quick reset before full reset.
					How long until the next failure requiring full reset?

2021-04-02.09	Enh: Limit potential out of control /etc/log/fping.txt.log.
					-> need to use the logfile rotation to do this, not now

2021-03-30.04	Enh: rename "quick" and "slow" web server options.
					-> minimal and full? not really. quick is ok. ...

2021-03-26.03   Enh: change messages log file location to /etc/log for persistence over a reboot.
                    Alt: copy messages* after a full reset.

2020.09.17.10   Enh: Sanity check for configured options.
                    -> do not specify -w quick, causes rapid loop, hard to break.

2020.09.17.09   Enh: Improve CheckVal to at least reject -options as values.



History:

0.8.7.1a 12/02/2022

2022-12-02.03	Enh: set default quick method to 'none' prep chip reload.

2022-12-02.01	Enh: make scanctrlflags configurable.
				-> method to configure on the fly using /etc/wlanpoke/scanctrl.inc:
				-> SCANOPT="scanctrl.inc"	# include file containing 6 values for wmiconfig --scan --scanctrlflags x x x x x x  plus other options
				-> SCANOPTDEF="0 0 0 0 0 0"	# default options for SCANOPT are upon launch, written to a new file if the file does not exist.
				-> local SCANOPTS=$(cat $SCANOPT) reads file in AR6002_reduceScans(). 
					if < 11 characters, the call is not made. 
					Otherwise the contents are appended to the wmiconfig call. The flags must come first. Can include other parameters.

0.8.7.0 11/30/2022

2022-11-30.01	Enh: Try to reduce number of access point scans to reduce failure
				-> New AR6002_reduceScans() calls 
					/lib/atheros/wmiconfig --scan --scanctrlflags 0 0 0 0 0 0 
				-> add AR6002_reduceScans() to ResetQuick() and RestartNetwork()

0.8.6.0 8/13/2022

2021-11-21.01	Bug: script exits with shell error when Access point has apostrophe
				-> add filter to CB_put () to filter all non-alpha characters from AP names.

2021-05-13.01   Doc: remove old or obsolete comments.
				-> Partially done


0.8.5.7 2021-05-19

2021-04-15.05	Enh: better interface to test different quick reset methods.
					-Qm option to specify quick reset method

2021-04-28.01	Doc: add warning to manual.txt to prevent ncat freeze on inadvertent mouse click
					-> Copied and introduced section from https://forums.slimdevices.com/showthread.php?109953-WiFi-connection-unstable-lost-on-three-Radios&p=1021974&viewfull=1#post1021974

2021-04-19.01	Bug: awstats defaults to /var/log/ if no command line option is used, missing modified script changes.
					-> If no option, parse the script to get the default LOGDIR is (might have been changed, and not an option)


0.8.5.6 4/30/2021	(experimental)

2021-04-28.03	Bug: Gaps and Resets list can grow to become quite large.
					-> Keep list trimmed. while do ... done was if then..fi 2 places.
					-> GapsListMax=240 and ResetsListMax=240 were both 120

2021-04-28.02	Enh: try power down and power up to unstick loss of sensitivity and see more APs during scan.
					-> new Quick_method=power  cycle power to mac and phy, etc. to try to reset radio sensitivity and
					PowerNap=500000     number of uSecs to sleep between disable and enable radio.
					-> that works to restore scan sensitivity, but ineffective for other issues.

0.8.5.5 4/18/2021	(experimental)

				Note: this version enables quick reset by default.
					-> PINGQUICK=3

2021-04-15.04	Enh: some sort of interface to test different quick reset methods.
					#Quick_method=wpa       # set to a quick reset mode, under development. (0.8.5.0)
                    #Quick_method=ifdnup76  # Not any better. For docmentation only.
					-- none of this is acceptable. See 2021-04-15.05


ahttpd.sh
#!/bin/sh
# Simple web server for wlan and 'wlanpoke' statistics
# Copyright (C) 2021 POMdev
# Derived from several web sources and wlanpoke,
#   see "https://stackoverflow.com/questions/16640054/minimal-web-server-using-netcat"
#
# This program is free software under GPL3 as stated in gpl3.txt, included.

Version="0.1.1 3/16/2021"

QUICKSVR="yes"          # set yes to enable periodic transmission of link statistics (0.5.0)
TCPPORT="8080"          # 1121 is default
APPDIR=$(dirname "$0")  # where the app files are located...

Help() {
    echo
    echo "Usage: $0 [options]"
    echo
    echo " -p * http server port (default $TCPPORT)"
    echo " -F   run slower 'full server' instead of default quick server (default NOT $QUICKSVR)"
    echo " -c   show copyright and license notice and exit"
    echo " -h   help and version ($Version)"
    echo
    echo "    * requires a reasonable value: no special chars, options are not validated."
    echo
    exit 0
}


CheckVal() {
    if [[ -z "$2" ]] ; then
      echo "$1" "requires a value"
      exit 1
    fi
}

while [ "$#" -ne 0 ]
do
case $1 in
        -p )    CheckVal $1 $2  ; shift ;   TCPPORT="$1"    ;;
        -F )    QUICKSVR="no"                       ;;
        -c )    cat "$APPDIR"/gpl3.txt  ;   exit 0  ;;
        -h )    Help            ;;
        * )     echo "Unsupported argument: '"$1"'" ;   exit 1
    esac
    shift
done

echo "$0 $Version" `pwd`    # 0.1.1: added pwd
cd "$APPDIR"
#pwd

# kill any previous instance (also kill any child processes)
PIDFILE="/var/run/ahttpd.pid"     # kill or be killed, Name hard coded.

KillApp () {
  if [ -r "$PIDFILE" ] ;  then  # 0.1.1 missing '$', how did that happen?
    PID=`cat "$PIDFILE"`
    echo "Killing process $PID"
    #kill -TERM '-'$PID     # DOESN'T WORK: kill child processes, too
    kill -TERM $PID         # 0.1.1 busybox CANNOT kill child processes, too
    kill -0 $PID >/dev/null 2>&1
    rm "$PIDFILE"
  fi
}

# we are running now
KillApp                         # kill any app still running
killall nc                      # stopgap: kill any rogue nc 'servers' (sorry wlanpoke if we stepped on you)
killall cat                     # 0.1.1: stopgap: kill any rogue cat pipe users
export WLPIPENAME="wlpipe"
rm -f $WLPIPENAME

# we are the new app process
PID=$$                          # not $!
echo "$PID" > $PIDFILE

if [[ "$QUICKSVR" == "yes" ]] ; then
  while true; do ./awstats.sh | nc -lp $TCPPORT; done
else
  # potential 'full service' web server uses 3 scripts, 4 processes: this main controller, a request parser, and one (or more) content generators.
  mkfifo $WLPIPENAME
  trap "rm -f $WLPIPENAME" EXIT
  while true
  do
    cat $WLPIPENAME | nc -lp $TCPPORT | ./arequest.sh
  done
fi


arequest.sh
#!/bin/sh
# Request parser for simple web server for wlan and 'wlanpoke' statistics
# Copyright (C) 2021 POMdev
# Derived from several web sources and wlanpoke, 
#   see "https://stackoverflow.com/questions/16640054/minimal-web-server-using-netcat"
#
# This program is free software under GPL3 as stated in gpl3.txt, included.

Version="0.1.1 3/16/2021"

#read stdin, launch script to extract then process request, and send it to a pipe, to be sent back to the browser.
export REQUEST=
while read -r line
do
  #line=$(echo "$line" | tr -d '\r\n')          # 0.1.1 do we really need this?
  # handle only simple GET requests, add code below for PUT processing.
  if echo "$line" | grep -qE '^GET /'           # if line starts with "GET /"
  then
    REQUEST=$(echo "$line" | cut -d ' ' -f2)    # extract the request
    # call a script here Note: REQUEST is exported, so the script can parse it (to answer 200/403/404 status code + content)
    ./awstats.sh > $WLPIPENAME                  # why wait?
  elif [ -z "$line" ]                           # empty line / end of request
  then
    #./awstats.sh > $WLPIPENAME
    echo "$REQUEST"                             # just to see how long it took for debugging
  fi
done



awstats.sh
#!/bin/sh
# Web page generator simple web server for wlan and 'wlanpoke' statistics
# Copyright (C) 2021 POMdev
# Derived from several web sources and wlanpoke,
#   see "https://stackoverflow.com/questions/16640054/minimal-web-server-using-netcat"
#
# This program is free software under GPL3 as stated in gpl3.txt, included.

Version="0.1.7 4/7/2021"

psEntry=`ps ax | grep "wlanpoke.sh -" | grep -v grep`
CmdLine=${psEntry##*/sh }
#/etc/wlanpoke/wlanpoke.sh -W slow -d /etc/log/
Options=${psEntry##*.sh }
#-W slow -d /etc/log/

#LOGDIR=`ps ax | grep "wlanpoke.sh -" | grep "\-d" | sed 's/.*-d \([^ ]*\).*/\1/g'`
#LOGDIR=`echo $Options | sed 's/.*-d \([^ ]*\).*/\1/g'`
# 0.1.7: | grep "\-d" restored, inadvertently deleted
LOGDIR=`echo $Options | grep "\-d" | sed 's/.*-d \([^ ]*\).*/\1/g'`
if [[ -z "$LOGDIR" ]] ; then
  LOGDIR="/var/log/"
fi


VersWLP="unknown $Options"
if [[ -r "Version" ]] ; then
  VersWLP=`cat "Version"`
fi



HOSTNAME=`cat "Hostname"`
if [[ -z "$HOSTNAME" ]] ; then
  HOSTNAME="WLAN"
fi


trq=`echo ${REQUEST##/}`        # get rid of root '/'
if [[ "$trq" == 'date' ]] ; then
  date
elif [[ -r "$trq" ]] ; then     # serve the file if it exists. To ... with the consequences!
  cat $trq
elif [[ "$trq" == 'RawFails' ]] ; then   # raw output for automated reports
  #echo $VersWLP "logs:" $LOGDIR
  echo $VersWLP
  echo `date -Iseconds` "(" `uptime` ")"
  cat ${LOGDIR}fping.txt
else
  echo -e "HTTP/1.1 200 OK\r"
  echo "Content-type: text/html"
  echo
  echo
  echo -e "<html><head><title>$HOSTNAME Statistics</title></head><body><h2>Wireless LAN Statistics</h2>\r\n"
  echo "<pre>"
  echo $VersWLP "logging to" $LOGDIR
  echo `date` "(" `uptime` ")"
  echo
  iwconfig eth1
  echo "</pre><h4>Ping: settings, events, and failed pings [0..n]; Step: full reset results, status; Gaps, Resets: # and recents</h4><pre>"
  cat ${LOGDIR}fping.txt
  echo "</pre><h4>recent AP scan results</h4><pre>"
  wpa_cli scan_results
  echo "</pre><h4>ar6002 chip statistics</h4><pre>"
  /lib/atheros/wmiconfig --getTargetStats
  echo "</pre><h4>important processes</h4><pre>"
  ps xf -o "pid,ppid,sess,pmem,vsz,size,rss,ni,pri,state,pcpu,start,time,args" | grep '[^]]$'       # 0.1.3 stat can show S<s which starts strikeout!!!
  echo -e "</pre>\r\n</body></html>"
fi
